src/fetchers/tarball-fetcher.js

```diff
@@ -65,58 +67,56 @@ export default class TarballFetcher extends BaseFetcher {
     reject: (error: Error) => void,
     tarballPath?: string,
   ): {
-    validateStream: crypto.HashStream,
+    validateStream: ssri.integrityStream,
     extractorStream: stream.Transform,
   } {
-    const validateStream = new crypto.HashStream();
+    let supportedIntegrity;
+    try {
+      supportedIntegrity = this._supportedIntegrity();
+    } catch (e) {
+      if (this.config.updateChecksums) {
+        // it's okay that we do not have any supported integrity, because we need to update it!
+        supportedIntegrity = {integrity: null, algorithms: ['sha1']};
+      } else {
+        reject(
+          new SecurityError(
```

Comment:
I think we should throw this `SecurityError` rather than calling `reject` - `reject` is supposed to be useful for async code only (like in the `finish` callback). It doesn't make sense to report an error but still return something.

Reply:
In theory I completely agree. The problem is that this stream pipeline is created inside a `Promise` (eg. https://github.com/imsnif/yarn/blob/a9e7ef26ea2ab534a160a2fbdf892a02368bdd24/src/fetchers/tarball-fetcher.js#L142). Since this is a synchronous error (as you say, we have enough information to throw this error before we return the `Promise`), it might be better to check it and throw appropriately before creating the pipeline using the `createExtractor` method (or better yet, create the pipeline before returning the `Promise` and then throw to our heart's content).
I haven't experimented with this, but my initial gut feeling tells me this would either involve duplicating a lot of logic (since we also have integrity checks that we need to perform and reject inside the pipeline) or a not-so-small refactoring of the tarball fetcher. I am very much in favour of the latter and might even volunteer for it if you'd like, but I feel it might be out of the scope of this PR. I hope you agree?
As for the former option (duplicating logic), I have a feeling it will make the code less coherent. I think the current solution, while not ideal, is the lesser evil at the moment... what do you think?

Reply:
I'm not sure I understand - if we throw instead of rejecting, since we are in a promise constructor, the promise will be automatically rejected. The only difference I can see would be that [these lines](https://github.com/imsnif/yarn/blob/a9e7ef26ea2ab534a160a2fbdf892a02368bdd24/src/fetchers/tarball-fetcher.js#L144-L146) would not be executed. Am I missing something?

Reply:
This was my misreading of the code, you are of course correct! For some reason I thought there were some async things before it and that the pipeline was being passed around. I tested it locally with throwing and it looks great.
Just pushed a fix and so far the CI seems happy.